relayer/chain/relaychain/connection.go

```diff
@@ -104,102 +104,79 @@ func (co *Connection) GetMMRLeafForBlock(
-	none := types.NewOptionU32Empty()
-	encoded, err := types.EncodeToBytes(none)
-	if err != nil {
-		co.log.WithError(err).Error("Error")
-		return nil, 0, types.Header{}, err
-	}
+type Head struct {
+	LeafIndex  int // order in which this head was returned from the storage query
+	ParaID uint32
+	Data   []byte
+}
 
-	baseParaHeadsStorageKey, err := types.CreateStorageKey(
-		co.GetMetadata(),
-		"Paras",
-		"Heads", encoded, nil)
-	if err != nil {
-		co.log.WithError(err).Error("Failed to create parachain header storage key")
-		return nil, 0, types.Header{}, err
-	}
+func (co *Connection) FetchParaHeads(blockHash types.Hash) (map[uint32]Head, error) {
 
-	//TODO fix this manual slice.
-	// The above types.CreateStorageKey does not give the same base key as polkadotjs needs for getKeys.
-	// It has some extra bytes.
-	// maybe from the none u32 in golang being wrong, or maybe slightly off CreateStorageKey call? we slice it
-	// here as a hack.
-	actualBaseParaHeadsStorageKey := baseParaHeadsStorageKey[:32]
-	co.log.WithField("actualBaseParaHeadsStorageKey", actualBaseParaHeadsStorageKey.Hex()).Info("actualBaseParaHeadsStorageKey")
+	keyPrefix := types.CreateStorageKeyPrefix("Paras", "Heads")
 
-	keysResponse, err := co.GetAPI().RPC.State.GetKeys(actualBaseParaHeadsStorageKey, blockHash)
+	keys, err := co.GetAPI().RPC.State.GetKeys(keyPrefix, blockHash)
 	if err != nil {
 		co.log.WithError(err).Error("Failed to get all parachain keys")
-		return nil, 0, types.Header{}, err
+		return nil, err
 	}
 
-	headersResponse, err := co.GetAPI().RPC.State.QueryStorage(keysResponse, blockHash, blockHash)
+	changeSets, err := co.GetAPI().RPC.State.QueryStorageAt(keys, blockHash)
 	if err != nil {
 		co.log.WithError(err).Error("Failed to get all parachain headers")
-		return nil, 0, types.Header{}, err
+		return nil, err
 	}
 
-	co.log.Info("Got all parachain headers")
-	var headers []types.Bytes
-	var ownParachainHeaderPos int
-	for _, headerResponse := range headersResponse {
-		for index, change := range headerResponse.Changes {
-
-			// TODO fix this manual slice with a proper type decode. only the last few bytes are for the ParaId,
-			// not sure what the early ones are for.
-			key := change.StorageKey[40:]
-			var parachainID types.U32
-			if err := types.DecodeFromBytes(key, &parachainID); err != nil {
-				co.log.WithError(err).Error("Failed to decode parachain ID")
-				return nil, 0, types.Header{}, err
+	heads := make(map[uint32]Head)
+
+	for _, changeSet := range changeSets {
+		for index, change := range changeSet.Changes {
+			if change.StorageData.IsNone() {
+				continue
 			}
 
-			var headerBytes types.Bytes
-			if err := types.DecodeFromBytes(change.StorageData, &headerBytes); err != nil {
-				co.log.WithError(err).Error("Failed to decode MMREncodableOpaqueLeaf")
-				return nil, 0, types.Header{}, err
+			var paraID uint32
+
+			if err := types.DecodeFromBytes(change.StorageKey[40:], &paraID); err != nil {
```

Comment:
any idea the last 40 bytes are the paraID/what the full type of the StorageKey actually is?

Reply:
The full key type is this:
```
ParaId: u32
Key: hash_twox_128("Paras") + hash_twox_128("Heads") + hash_twox_64(ParaId) + ParaId
```
The length of the first 3 hashes is 40 bytes which lines up with what our code is doing.

This article explains it well:
https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/